op-program/client/l2/engine.go

```diff
@@ -70,6 +76,10 @@ func (o *OracleEngine) GetPayload(ctx context.Context, payloadInfo eth.PayloadIn
 }
 
 func (o *OracleEngine) ForkchoiceUpdate(ctx context.Context, state *eth.ForkchoiceState, attr *eth.PayloadAttributes) (*eth.ForkchoiceUpdatedResult, error) {
+	header := o.backend.GetHeaderByHash(state.SafeBlockHash)
```

Comment:
curious why `SafeBlockHash` here? shouldn't this be the block that we're trying to update here?

Reply:
All 3 fork choice properties are set to the same value which is the block just before the one we're trying to execute. The block we're trying to update is the block after this one.

See: https://github.com/meyer9/optimism/blob/b9a93077954a832d1721a7ac37b470d0d2d06762/op-node/rollup/engine/engine_update.go#L85

Not sure why new payload isn't called before the block processing start, but seems like `InsertUnsafePayload` isn't called by `op-program` before block processing begins.

Reply:
This isn't always true. When op-program starts, the finalized, safe and unsafe heads are all set to the agreed block.  Since we load batches from L1 we have to use the block creation process to create the actual block from the transaction list which means sending `engine_forkChoiceUpdated` with a `PayloadAttributes`.  That's why we're hitting this before any calls to `NewPayload` - we have to build the block before we can import it. 

There's no `InsertUnsafePayload` prior to block processing because op-program doesn't receive any blocks via gossip. All inserted blocks come from L1 batches.

It's possible that op-program needs to import multiple blocks in a single run (span batches pre-Holocene) and in that case the unsafe head will have updated but the safe head may not have.  It's the `HeadBlockHash` that we will build the block on so this should be:
```suggestion
	header := o.backend.GetHeaderByHash(state.HeadBlockHash)
```

Reply:
Yeah, not sure what I was thinking - even though they're the same at the start, they're not guaranteed to always be the same. Updated this now [83d2a76](https://github.com/ethereum-optimism/optimism/pull/12559/commits/83d2a769bb7d43fbd0eb65854508080b52bc1bbf)